filemanager/filemanager.go

```diff
@@ -0,0 +1,120 @@
+//go:generate mockgen -destination=mock_filemanager/mock_filemanager.go -package mock_filemanager github.com/rudderlabs/rudder-go-kit/filemanager FileManager
+package filemanager
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+)
+
+const defaultTimeout = 120 * time.Second
+
+var (
+	ErrKeyNotFound            = errors.New("NoSuchKey")
+	ErrInvalidServiceProvider = errors.New("service provider not supported")
+)
+
+// Factory is a function that returns a new file manager
+type Factory func(settings *Settings) (FileManager, error)
+
+// UploadedFile contains information about the uploaded file
+type UploadedFile struct {
+	Location   string
+	ObjectName string
+}
+
+// FileInfo contains information about a file
+type FileInfo struct {
+	Key          string
+	LastModified time.Time
+}
+
+// FileManager is able to manage files in a storage provider
+type FileManager interface {
+	// ListFilesWithPrefix starts a list session for files with given prefix
+	ListFilesWithPrefix(ctx context.Context, startAfter, prefix string, maxItems int64) ListSession
+	// Download downloads the file with given key to the passed in file
+	Download(context.Context, *os.File, string) error
+	// Upload uploads the passed in file to the file manager
+	Upload(context.Context, *os.File, ...string) (UploadedFile, error)
+	// Delete deletes the file(s) with given key(s)
+	Delete(ctx context.Context, keys []string) error
+
+	// Prefix returns the prefix for the file manager
+	Prefix() string
+	// SetTimeout overrides the default timeout for the file manager
+	SetTimeout(timeout time.Duration)
+
+	// GetObjectNameFromLocation gets the object name/key name from the object location url
+	GetObjectNameFromLocation(string) (string, error)
+	// GetDownloadKeyFromFileLocation gets the download key from the object location url
+	GetDownloadKeyFromFileLocation(string) string
+}
+
+type ListSession interface {
+	// List returns next batch of files, until there are no more files
+	Next() (fileObjects []*FileInfo, err error)
+}
+
+// Settings for file manager
+type Settings struct {
+	Provider string
+	Config   map[string]interface{}
+	Logger   logger.Logger
+	Conf     *config.Config
+}
+
+// New returns file manager backed by configured provider
+func New(settings *Settings) (FileManager, error) {
+	log := settings.Logger
+	if log == nil {
+		log = logger.NewLogger().Child("filemanager")
```

Comment:
If log is nil, noop logger maybe might make more sense for the casual user? :thinking: